vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,615 @@
+}
+
+// TODO: This impl is a quick and dirty hack to allow the tests to continue using
+// `GuestMemoryMmap<()>`. Sadly this is exposed in the public API, but it should
+// be moved to an internal mock library.
+impl BitmapReplace for () {
+    type InnerBitmap = ();
+
+    // this implementation must not be used if the backend sets `VHOST_USER_PROTOCOL_F_LOG_SHMFD`
+    fn replace(&self, _bitmap: ()) {
+        panic!("The unit bitmap () must not be used if VHOST_USER_PROTOCOL_F_LOG_SHMFD is set");
+    }
+}
+
+impl MemRegionBitmap for () {
+    fn new<R: GuestMemoryRegion>(_region: &R, _logmem: Arc<MmapLogReg>) -> io::Result<Self> {
+        Err(io::Error::from(io::ErrorKind::Unsupported))
+    }
+}
+
+/// `BitmapMmapRegion` implements a bitmap tha can be replaced at runtime.
+/// The main use case is to support live migration on vhost-user backends
+/// (see `VHOST_USER_PROTOCOL_F_LOG_SHMFD` and `VHOST_USER_SET_LOG_BASE` in the vhost-user protocol
+/// specification). It uses a fixed memory page size of `VHOST_LOG_PAGE` bytes (i.e., `4096` bytes),
+/// so it converts addresses to page numbers before setting or clearing the bits.
+///
+/// To use this bitmap you need to define the memory as `GuestMemoryMmap<BitmapMmapRegion>`.
+///
+/// Note:
+/// This implementation uses `std::sync::RwLock`, the priority policy of the lock is dependent on
+/// the underlying operating system's implementation and does not guarantee any particular policy,
+/// in systems other than linux a thread trying to acquire the lock may starve.
+#[derive(Default, Debug, Clone)]
+pub struct BitmapMmapRegion {
+    // TODO: To avoid both reader and writer starvation we can replace the `std::sync::RwLock` with
+    // `parking_lot::RwLock`.
+    inner: Arc<RwLock<Option<AtomicBitmapMmap>>>,
+    base_address: usize, // The slice's base address
+}
+
+impl Bitmap for BitmapMmapRegion {
+    fn mark_dirty(&self, offset: usize, len: usize) {
+        let inner = self.inner.read().unwrap();
+        if let Some(bitmap) = inner.as_ref() {
+            if let Some(absolute_offset) = self.base_address.checked_add(offset) {
+                bitmap.mark_dirty(absolute_offset, len);
+            }
+        }
+    }
+
+    fn dirty_at(&self, offset: usize) -> bool {
+        let inner = self.inner.read().unwrap();
+        inner
+            .as_ref()
+            .is_some_and(|bitmap| bitmap.dirty_at(self.base_address.saturating_add(offset)))
+    }
+
+    fn slice_at(&self, offset: usize) -> <Self as WithBitmapSlice>::S {
+        Self {
+            inner: Arc::clone(&self.inner),
+            base_address: self.base_address.saturating_add(offset),
+        }
+    }
+}
+
+impl BitmapReplace for BitmapMmapRegion {
+    type InnerBitmap = AtomicBitmapMmap;
+
+    fn replace(&self, bitmap: AtomicBitmapMmap) {
+        let mut inner = self.inner.write().unwrap();
+        inner.replace(bitmap);
+    }
```

Comment:
I wonder: Is there a reason why we need this replace instead of just having a Bitmap from the start that supports the tracking but only enables it once we turn on logging?

That would not require all the new traits and just provide a vhost-specific Bitmap implementation for the ordinary Bitmap trait.

Or am I missing something that this trait definition does not allow that we need?

Reply:
> I wonder: Is there a reason why we need this replace instead of just having a Bitmap from the start that supports the tracking but only enables it once we turn on logging?

Well, that is what we do, from the beginning we start with `BitmapMmapRegion` (because the vm-memory crate requires the bitmap be defined at compile time), but we don't know where is the backing memory of that bitmap, we need to wait until the front-end sends the `FD`, that we need to mmap, as part of the `SET_LOG_BASE` message. So we start with an "empty" bitmap, and we replace is memory as soon as we have it.

We could just replace the memory instead of the InnerBitmap (that is the plan for the future) but due to the public API and how the current test expect a () bitmap, it makes all the changes more complicated/extensive (I plan to do it anyway).